src/linker/Linker/AssemblyResolver.cs

```diff
@@ -68,12 +72,57 @@ public AssemblyResolver (Dictionary<string, AssemblyDefinition> assembly_cache)
 			set { _context = value; }
 		}
 
+#if !FEATURE_ILLINK
+		// The base class's definition of GetAssembly is visible when using DirectoryAssemblyResolver.
+		AssemblyDefinition GetAssembly (string file, ReaderParameters parameters)
+		{
+			if (parameters.AssemblyResolver == null)
+				parameters.AssemblyResolver = this;
+
+			return ModuleDefinition.ReadModule (file, parameters).Assembly;
+		}
+#endif
+
+		private AssemblyDefinition ResolveWithPaths (AssemblyNameReference name, ReaderParameters parameters)
+		{
+			// Validate arguments, similarly to how the base class does it.
+			if (name == null)
+				throw new ArgumentNullException ("name");
+			if (parameters == null)
+				throw new ArgumentNullException ("parameters");
+
+			AssemblyDefinition asm = null;
+
+			// Try the new resolution behavior. This can't live in the cecil-owned base class.
+			asm = SearchAssemblyPaths (name, _assemblyPaths, parameters);
+			if (asm != null)
+				return asm;
+
+			// Fall back to the base class resolution logic
+			return base.Resolve (name, parameters);
+		}
+
+		private AssemblyDefinition SearchAssemblyPaths (AssemblyNameReference name, IEnumerable<string> assemblyPaths, ReaderParameters parameters)
```

Comment:
Why base implementation cannot be used?

Reply:
The base classes currently only have logic to search in directories (BaseAssemblyResolver from cecil, or DirectoryAssemblyResolver from the linker). To move it to the base class, we would need to do one of:
1. Modify cecil's resolution behavior. I've attempted this in the past, but Jb suggested that the linker should own any resolution logic specific to its purposes.
2. Move it to DirectoryAssemblyResolver but not BaseAssemblyResolver, and support `--ref` only on illink, not on monolinker.
3. Keep it in the child class, extending both base classes.

I opted for the third approach, to keep the command-line consistent between illink/monolinker (I thought this option may be useful for monolinker as well, since it could in theory hit similar issues).

If you prefer, I can also make the change only for illink, and move the change to DirectoryAssemblyResolver only. This is actually what I did initially (see the first commit), before I realized that monolinker wouldn't support the change this way.

Reply:
Please rename it the to `ResolveFromReferences (AssemblyNameReference name, Collection<string> references, ReaderParameters parameters)`